src/librustc_mir/interpret/validity.rs

```diff
@@ -24,43 +25,41 @@ use super::{
 };
 
 macro_rules! throw_validation_failure {
-    ($what:expr, $where:expr, $details:expr) => {{
-        let mut msg = format!("encountered {}", $what);
-        let where_ = &$where;
-        if !where_.is_empty() {
-            msg.push_str(" at ");
-            write_path(&mut msg, where_);
-        }
-        write!(&mut msg, ", but expected {}", $details).unwrap();
-        throw_ub!(ValidationFailure(msg))
-    }};
-    ($what:expr, $where:expr) => {{
+    ($what:expr, $where:expr $(, $details:expr )?) => {{
         let mut msg = format!("encountered {}", $what);
         let where_ = &$where;
         if !where_.is_empty() {
             msg.push_str(" at ");
             write_path(&mut msg, where_);
         }
+        $( write!(&mut msg, ", but expected {}", $details).unwrap(); )?
         throw_ub!(ValidationFailure(msg))
     }};
 }
 
+/// Returns a validation failure for any Err value of $e.
 macro_rules! try_validation {
-    ($e:expr, $what:expr, $where:expr, $details:expr) => {{
-        match $e {
-            Ok(x) => x,
-            // We re-throw the error, so we are okay with allocation:
-            // this can only slow down builds that fail anyway.
-            Err(_) => throw_validation_failure!($what, $where, $details),
-        }
+    ($e:expr, $what:expr, $where:expr $(, $details:expr )?) => {{
+        try_validation_pat!($e, _, $what, $where $(, $details )?)
     }};
-
-    ($e:expr, $what:expr, $where:expr) => {{
+}
+/// Like try_validation, but will throw a validation error if any of the patterns in $p are
+/// matched. Other errors are passed back to the caller, unchanged. This lets you use the patterns
+/// as a kind of validation blacklist:
+///
+/// ```rust
+/// let v = try_validation_pat(some_fn(), Foo | Bar | Baz, "some failure", "some place");
+/// // Failures that match $p are thrown up as validation errors, but other errors are passed back
+/// // unchanged.
+/// ```
+macro_rules! try_validation_pat {
+    ($e:expr, $( $p:pat )|*, $what:expr, $where:expr $(, $details:expr )?) => {{
```

Comment:
I wonder if we can come up with a better syntax that this function-like one? Maybe
```rust
try_validation_pat!(some_fn() {
  Foo | Bar | Baz => "some failure" @ self.path
})
```
That's more readable -- maybe?

Reply:
I'd 100% be down for something like this, but the issue is that `expr`s can only be followed by `,`, `=>` or `;`, so it'd have to be something like

```rust
try_validation_pat!(some_fn(); {
    Foo | Bar | Baz => "some failure", details, @ self.path,
    Quux | Quuux => "some other failure", @ self.path,
});
```
ie, you'd need to delimit `some_fn()`, and you'd also need to delimit the what, where and details. The sigil `@` on the "where" parameter would also be necessary, so that the end of each case can be distinguished. An alternative would be to have each arm delimited by a `;`, but then it wouldn't look similar to a match expression anymore.

Hmm, looking at it now, that's still a fair readability improvement over the function-like syntax . It's definitely easier to tell which parameters are what. Downside is that it's as natural to write as the syntax you suggested. Would you mind this syntax?

Reply:
Can't we use custom tokens (like `@path`) to delimit expressions?

Reply:
Hm, looks like we cannot, bummer. So let's improvise. Based on what you proposed... but also, `self.path` will be the same for all branches... so what about this?
```rust
try_validation_pat!(some_fn(), self.path, {
    Foo | Bar | Baz => "bad bytes", @expected "a pointer",
    Quux | Quuux => "some other failure",
});
```
Ideally we could even have built-in support for format strings, instead of repeating `format_args!` everywhere. Maybe `what` and `details` (should be renamed to `expected`?) should just be `tt` and we add some extra parentheses?